syncer/filter_test.go

```diff
@@ -91,41 +71,129 @@ func (s *testFilterSuite) TestSkipQueryEvent(c *C) {
 		},
+	}
+	p := parser.New()
+	for _, ca := range cases {
+		stmt, err := p.ParseOneStmt(ca.sql, "", "")
+		c.Assert(err, IsNil)
+		skipped, err2 := syncer.skipQueryEvent(ca.tables, stmt, ca.sql)
+		c.Assert(err2, IsNil)
+		c.Assert(skipped, Equals, ca.expectSkipped)
+	}
+}
+
+func (s *testFilterSuite) TestSkipRowsEvent(c *C) {
+	syncer := &Syncer{}
+	filterRules := []*bf.BinlogEventRule{
 		{
-			"create table tx.test (id int)",
-			[]*filter.Table{{Schema: "tx", Name: "test"}},
-			false,
+			SchemaPattern: "foo*",
+			TablePattern:  "",
+			Events:        []bf.EventType{bf.InsertEvent},
+			SQLPattern:    []string{""},
+			Action:        bf.Ignore,
 		},
-		// test schema rule
+	}
+	var err error
+	syncer.binlogFilter, err = bf.NewBinlogEvent(false, filterRules)
+	c.Assert(err, IsNil)
+
+	cases := []struct {
+		table     *filter.Table
+		eventType replication.EventType
+		expected  bool
+	}{
 		{
-			"create table foo.test(id int)",
-			[]*filter.Table{{Schema: "foo", Name: "test"}},
+			// test filter one event
+			&filter.Table{Schema: "foo", Name: "test"},
+			replication.WRITE_ROWS_EVENTv0,
+			true,
+		}, {
+			&filter.Table{Schema: "foo", Name: "test"},
+			replication.UPDATE_ROWS_EVENTv0,
+			false,
+		}, {
+			&filter.Table{Schema: "foo", Name: "test"},
+			replication.DELETE_ROWS_EVENTv0,
 			false,
 		},
+	}
+	for _, ca := range cases {
+		needSkip, err2 := syncer.skipRowsEvent(ca.table, ca.eventType)
+		c.Assert(err2, IsNil)
+		c.Assert(needSkip, Equals, ca.expected)
+	}
+}
+
+func (s *testFilterSuite) TestFilterOneEvent(c *C) {
+	cfg := &config.SubTaskConfig{
+		BAList: &filter.Rules{
+			IgnoreDBs: []string{"s1"},
+		},
+	}
+	syncer := NewSyncer(cfg, nil)
+	// test binlog filter
+	filterRules := []*bf.BinlogEventRule{
 		{
-			"rename table foo.test to foo.test1",
-			[]*filter.Table{{Schema: "foo", Name: "test"}},
-			true,
+			SchemaPattern: "*",
+			TablePattern:  "",
+			Events:        []bf.EventType{bf.DropTable},
+			SQLPattern:    []string{"^drop\\s+table"},
+			Action:        bf.Ignore,
 		},
-		// test table rule
+	}
+	var err error
+	syncer.binlogFilter, err = bf.NewBinlogEvent(false, filterRules)
+	c.Assert(err, IsNil)
+
+	cases := []struct {
+		sql           string
+		table         *filter.Table
+		eventType     bf.EventType
+		expectSkipped bool
+	}{
 		{
-			"create table foo.bar(id int)",
-			[]*filter.Table{{Schema: "foo", Name: "bar"}},
+			// system table
+			"create table mysql.test (id int)",
+			&filter.Table{Schema: "mysql", Name: "test"},
+			"",
 			true,
+		}, {
+			// test binlog filter
+			"drop table tx.test",
+			&filter.Table{Schema: "tx", Name: "test"},
+			bf.DropTable,
+			true,
+		}, {
+			"create table tx.test (id int)",
+			&filter.Table{Schema: "tx", Name: "test"},
+			bf.CreateTable,
+			false,
+		}, {
+			// test balist
+			"create table s1.test (id int)",
+			&filter.Table{Schema: "s1", Name: "test"},
+			bf.CreateTable,
+			false, // there should be true, so what does BAList configure?
```

Comment:
PTAL @lance6716

Reply:
maybe baList is not initialized from config